### ISSUE TITLE
Reference conflicts

### ISSUE THREAD

**Issue by theofidry** [nelmio/alice#420]
For a given fixture set:

```yaml
Nelmio\Entity\User:
  user1: {}

Nelmio\Entity\Group:
  group1: {}
```

The array of objects you get from the loader will be:

```php
[
  'user1' => Nelmio\Entity\User#object,
  'group1' => Nelmio\Entity\Group#object,
]
```

This lead to a very simple problem, if you have:

```yaml
Nelmio\Entity\User:
  user1: {}

Nelmio\Entity\Group:
  user1: {} # same key as for Nelmio\Entity\User
```

Then you will get:

```php
[
  'user1' => Nelmio\Entity\Group#object,
]
```

As it is now, I'm not sure if we can do anything about it, we could potentially change it to have references relative to the class, e.g.:

```php
[
  'Nelmio\Entity\User#user1' => Nelmio\Entity\User#object,
  'Nelmio\Entity\Group#group1' => Nelmio\Entity\Group#object,
]
```

But that's potentially a big BC. Another solution is to keep it as it is, but then it would be nice to have a warning somewhere about this conflict... Also another thing worth mentioning is that the problem is greatly mitigated by the fact that you naturally tend to repeat a bit of the class name (e.g. `userX` for `Nelmio\Entity\User`, `groupX` for `Nelmio\Entity\Group`, etc.). But the problem is more likely to occur if you have the same class declared in different namespaces, e.g. `Nelmio\Entity\User` and `Nelmio\Bundle\FosUserBundle\Entity\User`).

@tshelburne WDYT?

**Comment by tshelburne** [comment]
I don't think we can / should introduce the relative reference naming - then the downstream user is required to follow those conventions *every* time they want to use a reference.

Having created a fairly large set of fixtures using Alice that ran into these exact problems, I think our responsibility stops at notifying the user with a warning when the naming match occurs, and then leaving it to them to decide how they would like to solve it (likely, simply by renaming one of the fixtures more specifically).

Basically, I wouldn't complicate things.

**Comment by theofidry** [comment]
Let's go for a deprecation notice in v2 to warn the user and an exception in v3 then.

**Comment by theofidry** [comment]
Actually turns out that this is much more tricky than expected. The only way I can see to fix this issue in 3.x would be to iterate over all the fixtures before creating any of them to detect such name conflicts. This can be easily done but I'm not sure if it's worth it, especially if the user loads one file after another: in case of conflicts, the existing object is overridden by the new one and this is a perfectly valid behaviour, I don't see the need of warning or anything there. As this solution also has a little performance cost, I just don't think it's worth it.

**Comment by tshelburne** [comment]
@theofidry Instead of iterating, why not just check for existence at the time when you add the new instance to the list?

**Comment by theofidry** [comment]
Because as fixture objects are immutable I re-create them several times, so
I have no easy way to tell if the previous one is the same one (no
conflict) or a new one (conflicts) :/
